qiskit/aqua/operators/primitive_ops/summed_pauli_op.py

```diff
@@ -0,0 +1,384 @@
+        """Permutes the sequence of ``SummedPauliOp``.
+
+        Args:
+            permutation: A list defining where each Pauli should be permuted. The Pauli at index
+                j of the primitive should be permuted to position permutation[j].
+
+        Returns:
+              A new SummedPauliOp representing the permuted operator. For operator (X ^ Y ^ Z) and
+              indices=[1,2,4], it returns (X ^ I ^ Y ^ Z ^ I).
+
+        Raises:
+            AquaError: if indices do not define a new index for each qubit.
+        """
+        if len(permutation) != self.num_qubits:
+            raise AquaError(
+                "List of indices to permute must have the same size as Pauli Operator"
+            )
+        length = max(permutation) + 1
+        spop = self.primitive.tensor(  # type:ignore
+            SparsePauliOp(Pauli(label="I" * (length - self.num_qubits)))
+        )
+        permutation = [i for i in range(length) if i not in permutation] + permutation
+        permutation = np.arange(length)[np.argsort(permutation)]
+        permutation = np.hstack([permutation, permutation + length])  # type: ignore
+        spop.table.array = spop.table.array[:, permutation]
+        return SummedPauliOp(spop, self.coeff)
+
+    def compose(
+            self,
+            other: OperatorBase,
+            permutation: Optional[List[int]] = None,
+            front: bool = False,
+    ) -> OperatorBase:
+
+        new_self, other = self._expand_shorter_operator_and_permute(other, permutation)
+        new_self = cast(SummedPauliOp, new_self)
+
+        if front:
+            return other.compose(new_self)
+        # If self is identity, just return other.
+        if not np.any(new_self.primitive.table.array):  # type: ignore
+            return other * new_self.coeff * sum(new_self.coeffs)  # type: ignore
+
+        # Both SummedPauliOps
+        if isinstance(other, SummedPauliOp):
+            return SummedPauliOp(
+                new_self.primitive * other.primitive,  # type:ignore
+                coeff=new_self.coeff * other.coeff,
+            )
+        # TODO: implement compose with PauliOp
+
+        # pylint: disable=cyclic-import,import-outside-toplevel
+        from ..state_fns.circuit_state_fn import CircuitStateFn
+        from .circuit_op import CircuitOp
+
+        if isinstance(other, (CircuitOp, CircuitStateFn)):
+            return new_self.to_circuit_op().compose(other)
+
+        return super(SummedPauliOp, new_self).compose(other)
+
+    def to_matrix(self, massive: bool = False) -> np.ndarray:
+        OperatorBase._check_massive("to_matrix", True, self.num_qubits, massive)
+        return self.primitive.to_matrix() * self.coeff  # type: ignore
+
+    def __str__(self) -> str:
+        prim_list = self.primitive.to_list()  # type:ignore
+
+        def format_number(x):
+            return x.real if np.isreal(x) else x
+
+        coeff_string = "" if self.coeff == 1 else f" * {self.coeff}"
+
+        return (
```

Comment:
This method looks more like a __repr__ than a __str__. Should it be changed to __repr__ which will be used anyway if some does a str conversion and there is not a more user friendly __str__ defined (maybe this is user friendly enough anyway).

Reply:
`repr` is evaluatable in many case. Official docs said

> If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment).

Therefore, eval(repr(SummedPauliOp)) is valid now.

This str is human readble.
str(SummedOp[PauliOp]) is like this
```
SummedOp([
  -1.052373245772859 * II,
  0.39793742484318045 * IZ,
  -0.39793742484318045 * ZI,
  -0.01128010425623538 * ZZ,
  0.18093119978423156 * XX
])
```
I made str of SummedPauliOp similarly.

Reply:
Ah ok, because its a PrimitiveOp it ends up inheriting repr from the bsae class. If that is how the other human readable ones are, that look like constructors, then I guess its consistent.

Reply:
At first I was thinking of making it like this.
```
- 1.052373245772859 * II
+ 0.39793742484318045 * IZ
- 0.39793742484318045 * ZI
- 0.01128010425623538 * ZZ
+ 0.18093119978423156 * XX
```
Actually, I just chose the easier one, so if this one is better, I'll change it.

Reply:
In looking the PrimitiveOps do not have that constructor syntax so having some simpler/more readable style being compatible with the formatting of those which show the coeff * primitive or just primitive  when coeff is 1 seems reasonable and consistent with it being a PrimitiveOp.